packages/core/src/validate/error.ts

```diff
@@ -0,0 +1,72 @@
+import chalk from 'chalk';
+
+import { ValidationError } from './run';
+import { UpgradesError, ErrorDescriptions } from '../error';
+
+export class ValidationErrors extends UpgradesError {
+  constructor(contractName: string, readonly errors: ValidationError[]) {
+    super(`Contract \`${contractName}\` is not upgrade safe`, () => {
+      return errors.map(describeError).join('\n\n');
+    });
+  }
+}
+
+const errorInfo: ErrorDescriptions<ValidationError> = {
+  constructor: {
+    msg: e => `Contract \`${e.contract}\` has a constructor`,
+    hint: () => 'Define an initializer instead',
+    link: 'https://zpl.in/upgrades/error-001',
+  },
+  delegatecall: {
+    msg: () => `Use of delegatecall is not allowed`,
+    link: 'https://zpl.in/upgrades/error-002',
+  },
+  selfdestruct: {
+    msg: () => `Use of selfdestruct is not allowed`,
+    link: 'https://zpl.in/upgrades/error-003',
+  },
+  'state-variable-assignment': {
+    msg: e => `Variable \`${e.name}\` is assigned an initial value`,
+    hint: () => 'Move the assignment to the initializer',
+    link: 'https://zpl.in/upgrades/error-004',
+  },
+  'state-variable-immutable': {
+    msg: e => `Variable \`${e.name}\` is immutable`,
+    hint: () => `Use a constant or mutable variable instead`,
+    link: 'https://zpl.in/upgrades/error-005',
+  },
+  'external-library-linking': {
+    msg: e => `Linking external libraries like \`${e.name}\` is not yet supported`,
+    hint: () =>
+      `Use libraries with internal functions only, or skip this check with the \`unsafeAllowLinkedLibraries\` flag \n` +
+      `    if you have manually checked that the libraries are upgrade safe`,
+    link: 'https://zpl.in/upgrades/error-006',
+  },
+  'struct-definition': {
```

Comment:
Are these (struct-definition & enum-definition) still needed ? Should they be removed ?

Reply:
These are needed because this will be a backwards compatible release, and in theory someone could be using an older Hardhat plugin with the newer Core. This could result in older cached validations to be passed to the core, and they can contain these errors. Those older versions of the Hardhat plugin will not invalidate the cache if it's an older version.

We could still detect it and throw an error, but I preferred to go this route for some reason. What do you think?